pkg/commands/export/component_export.go

```diff
@@ -0,0 +1,133 @@
+	log "github.com/golang/glog"
+	"github.com/spf13/cobra"
+)
+
+const defaultPerms = 0664
+
+// compOptions represents options flags for the export command.
+type compOptions struct {
+	// writeTo indicates the directory to write the components to. This is joined with the pa
+	writeTo string
+
+	// overwrite indicates whether to overwrite existing files.
+	overwrite bool
+
+	// pathTemplates are paths to generate. This is joined with the
+	// writeTo directory
+	pathTemplates []string
+
+	// exportSet determines whether to export a set for all the known components.
+	exportSet bool
+
+	// setName is an optional set name for the set of components.
+	setName string
+
+	// version is an optional version for the components
+	setVersion string
+}
+
+// compOpts is a global options instance for reference via the add commands.
+var compOpts = &compOptions{}
+
+func componentAction(ctx context.Context, fio files.FileReaderWriter, sio cmdlib.StdioReaderWriter, cmd *cobra.Command, _ []string) {
+	gopt := cmdlib.GlobalOptionsValues.Copy()
+	brw := cmdlib.NewBundleReaderWriter(fio, sio)
+	if err := runComponent(ctx, compOpts, brw, sio, fio, gopt); err != nil {
+		log.Exit(err)
+	}
+}
+
+func runComponent(ctx context.Context, o *compOptions, brw cmdlib.BundleReaderWriter, stdio cmdlib.StdioReaderWriter, fio files.FileReaderWriter, gopt *cmdlib.GlobalOptions) error {
+	bw, err := brw.ReadBundleData(ctx, gopt)
+	if err != nil {
+		return fmt.Errorf("error reading bundle contents: %v", err)
+	}
+
+	comps := bw.AllComponents()
+
+	var set *bundle.ComponentSet
+	if o.exportSet {
+		setName, setVersion := getSetNameVersion(bw, o)
+		set, err = converter.NewExporter(comps...).ComponentSet(setName, setVersion)
+		if err != nil {
+			return err
+		}
+	}
+
+	if o.writeTo != "" {
+		return writeToFiles(ctx, comps, o, fio, set)
+	} else {
+		return writeToStdout(comps, stdio, set)
+	}
+}
+
+func getSetNameVersion(bw *wrapper.BundleWrapper, o *compOptions) (string, string) {
+	setName := o.setName
+	setVersion := o.setVersion
+	if bw.Kind() == "bundle" {
+		bun := bw.Bundle()
+		setName = bun.SetName
+		setVersion = bun.Version
+	}
+	return setName, setVersion
+}
+
+func writeToFiles(ctx context.Context, comps []*bundle.Component, o *compOptions, fio files.FileReaderWriter, set *bundle.ComponentSet) error {
+	pathMap, err := converter.NewExporter(comps...).ComponentsWithPathTemplates(o.pathTemplates, set)
+	if err != nil {
+		return err
+	}
+	for cpath, yaml := range pathMap {
+		fpath := filepath.Join(o.writeTo, cpath)
+		dir := filepath.Dir(fpath)
+		// TODO(kashomon): This is going to be a problem for mocking.
+		if _, err := os.Stat(dir); os.IsNotExist(err) {
+			os.MkdirAll(dir, os.ModePerm)
+		}
+		if !o.overwrite {
+			if _, err := os.Stat(fpath); err == nil {
+				return fmt.Errorf("file %q already exists", fpath)
+			}
+		}
+		if err := fio.WriteFile(ctx, fpath, []byte(yaml), defaultPerms); err != nil {
```

Comment:
This should probably be atomic.

Reply:
What do you mean?

Reply:
In the overwrite mode, errors during write can lead to corruption (e.g. is only partially completed leaving half the file written). If this is a static pod manifest, this would break a previously good configuration. I think we want to try to commit a change like this atomically like we do for projected volume files:

https://github.com/kubernetes/kubernetes/blob/81e7858ece4713e043d61d96874481134c77b086/pkg/volume/util/atomic_writer.go#L72

Reply:
Do you feel like this is urgent and something I should do in this CL or it should be a followup? Or even an open issue that gets resolved at some point? 

It looks like making it atomic is a bit more work is why I'm asking

Reply:
No, not urgent

Reply:
Filed https://github.com/GoogleCloudPlatform/k8s-cluster-bundle/issues/204

Reply:
At least in my use case, there is no concurrency and any error during write would abort the whole process. So this does not seem urgent to me.